include/lbann/layers/transform/pooling.hpp

```diff
@@ -136,6 +135,8 @@ class pooling_layer : public data_type_layer<TensorDataType>
   /** Tensor DNN library descriptors. */
   dnn_lib::data_parallel_layer_tensor_manager<TensorDataType>
     m_tensors_dnn_desc;
+  /** Workspace (shared between forward and backprop by MIOpen). */
+  std::unique_ptr<AbsMatrixType> m_workspace;
```

Comment:
> But you have `TensorDataType` and `Dev`.
> 
> If we don't care about a few extra bytes, the `unique_ptr` also isn't necessary.

We do (and that's how I implemented it in the first round), but it doesn't compile because of the .reset() calls (`Copy()` returns an abstractmatrix)

Reply:
Well the `.reset` calls should be `= std::make_unique` calls. And the `->Copy` call isn't needed when you can just use the copy constructor `other.m_workspace ? std::make_unique<MatrixType>(*other.m_workspace) : nullptr`. So let's consider just moving the whole thing to an `El::Matrix<T, Device::GPU>` and dispense with this ptr nonsense. I'll give you a patch offline.